lox/ast.hpp

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// SPDX-FileCopyrightText: Copyright (c) assignUser
+
+#pragma once
+#include <memory>
+#include <cstddef>
+
+#include "lox/token.hpp"
+
+class Binary;
+class Grouping;
+class String;
+class Number;
+class Boolean;
+class Nil;
+class Unary;
+
+class Visitor {
+public:
+  virtual ~Visitor() = default;
+  virtual void visit(Binary const &expr) = 0;
+  virtual void visit(Grouping const &expr) = 0;
+  virtual void visit(String const &expr) = 0;
+  virtual void visit(Number const &expr) = 0;
+  virtual void visit(Boolean const &expr) = 0;
+  virtual void visit(Nil const &expr) = 0;
+  virtual void visit(Unary const &expr) = 0;
+};
+
+class Expr {
+public:
+  virtual ~Expr() = default;
+  virtual void accept(Visitor &visitor) const = 0;
+};
+
+class Binary : public Expr {
+public:
+  Binary(std::unique_ptr<Expr> lhs, Token op, std::unique_ptr<Expr> rhs): lhs{std::move(lhs)}, op{std::move(op)}, rhs{std::move(rhs)}{}
```

Comment:
Prefer making the constructors private if the intended acceptable usage is only via invocations to `::make(...)`